package/MDAnalysis/analysis/gnm.py

```diff
@@ -393,25 +406,32 @@ def generate_kirchoff(self):
         sqrt_res_sizes = np.sqrt([r.atoms.n_atoms for r in  self.ca.residues]) \
                          if self.weights == "size" else None
 
-        for icounter in range(natoms):
-            neighbour_atoms = []
-            for x in (-1, 0, 1):
-                gx = res_positions[icounter][0] + x
-                if 0 <= gx < len(grid):
-                    for y in (-1, 0, 1):
-                        gy = res_positions[icounter][1] + y
-                        if 0 <= gy < len(grid[0]):
-                            for z in (-1, 0, 1):
-                                gz = res_positions[icounter][2] + z
-                                if 0 <= gz < len(grid[0][0]):
-                                    neighbour_atoms += grid[gx][gy][gz]
-            for jcounter in neighbour_atoms:
-                if jcounter > icounter and _dsq(positions[icounter], positions[jcounter]) <= cutoffsq:
-                    iresidue, jresidue = residue_index_map[icounter], residue_index_map[jcounter]
-                    contact = 1.0 / (sqrt_res_sizes[iresidue] * sqrt_res_sizes[jresidue]) \
-                              if self.weights == "size" else 1.0
-                    matrix[iresidue][jresidue] -= contact
-                    matrix[jresidue][iresidue] -= contact
-                    matrix[iresidue][iresidue] += contact
-                    matrix[jresidue][jresidue] += contact
+
+        for cell_x in range(len(grid)):
+            for cell_y in range(len(grid[0])):
+                for cell_z in range(len(grid[0][0])):
+                    atoms = grid[cell_x][cell_y][cell_z]
+                    all_atoms = []
+                    for x in (-1, 0, 1):
+                        gx = cell_x + x
+                        if 0 <= gx < len(grid):
+                            for y in (-1, 0, 1):
+                                gy = cell_y + y
+                                if 0 <= gy < len(grid[0]):
+                                    for z in (-1, 0, 1):
+                                        gz = cell_z + z
+                                        if 0 <= gz < len(grid[0][0]):
+                                            all_atoms += grid[gx][gy][gz]
+
+                    for i_atom in atoms:
+                        for j_atom in all_atoms:
+                            if j_atom > i_atom and _dsq(positions[i_atom], positions[j_atom]) < cutoffsq:
+                                iresidue, jresidue = residue_index_map[i_atom], residue_index_map[j_atom]
+                                contact = 1.0 / (sqrt_res_sizes[iresidue] * sqrt_res_sizes[jresidue]) \
+                                        if self.weights == "size" else 1.0
+                                matrix[iresidue][jresidue] -= contact
+                                matrix[jresidue][iresidue] -= contact
+                                matrix[iresidue][iresidue] += contact
+                                matrix[jresidue][jresidue] += contact
+
```

Comment:
I've slightly changed the used algorithm here. This saves about a second on my laptop. This means we see maybe a 10-15% speed up.  In the previous algorithm we went through the list of atoms and searched for possible neighbours in a grid. Now I iterate over the grid and calculated all distance values for the atoms in a grid cell. This saves recomputing the list of candidate neighbours from the grid.